File: cmd/listen.go
package cmd

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/linq-team/linq-cli/internal/config"
	"github.com/linq-team/linq-cli/internal/apiv3"
	"github.com/spf13/cobra"
)

var listenCmd = &cobra.Command{
	Use:   "listen",
	Short: "Listen for incoming messages",
	Long: `Start a local server that receives webhook events from Linq.
Automatically registers a webhook subscription and streams events to stdout.`,
	Example: `  linq listen
  linq listen --port 4567
  linq listen --events message.received,message.delivered
  linq listen --json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		port, _ := cmd.Flags().GetInt("port")
		jsonOutput, _ := cmd.Flags().GetBool("json")

		cfg, err := config.Load()
		if err != nil {
			return fmt.Errorf("not authenticated. Run 'linq login' first")
		}

		token := resolveToken(cmd, cfg)
		if token == "" {
			return fmt.Errorf("no token found. Run 'linq login' or pass --token")
		}

		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		sigCh := make(chan os.Signal, 1)
		signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
		go func() {
			<-sigCh
			fmt.Fprintln(os.Stderr, "\nShutting down...")
			cancel()
		}()

		eventCh := make(chan json.RawMessage, 100)
		srv := startWebhookServer(port, eventCh)

		fmt.Fprintf(os.Stderr, "Listening for events on :%d\n", port)
		fmt.Fprintf(os.Stderr, "Press Ctrl+C to stop\n\n")

		go func() {
			for raw := range eventCh {
				if jsonOutput {
					fmt.Println(string(raw))
				} else {
					printEvent(raw)
				}
			}
		}()

		<-ctx.Done()

		shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer shutdownCancel()
		srv.Shutdown(shutdownCtx)

		return nil
	},
}

func startWebhookServer(port int, eventCh chan<- json.RawMessage) *http.Server {
	mux := http.NewServeMux()
	mux.HandleFunc("/webhook", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}

		body, err := io.ReadAll(r.Body)
		if err != nil {
			http.Error(w, "failed to read body", http.StatusBadRequest)
			return
		}

		eventCh <- json.RawMessage(body)
		w.WriteHeader(http.StatusOK)
	})

	srv := &http.Server{
		Addr:    fmt.Sprintf(":%d", port),
		Handler: mux,
	}

	go srv.ListenAndServe()
	return srv
}

func printEvent(raw json.RawMessage) {
	var peek struct {
		EventType string `json:"event_type"`
	}
	if err := json.Unmarshal(raw, &peek); err != nil {
		fmt.Fprintf(os.Stderr, "failed to parse event: %v\n", err)
		return
	}

	switch peek.EventType {
	case "message.received":
		var wh apiv3.MessageReceivedWebhook
		if err := json.Unmarshal(raw, &wh); err == nil {
			from := ""
			if wh.Data.From != nil {
				from = *wh.Data.From
			}
			msg := extractPayloadText(wh.Data.Message)
			fmt.Printf("[%s] << %s: %s\n", wh.CreatedAt.Format(time.RFC3339), from, msg)
		}

	case "message.sent":
		var wh apiv3.MessageSentWebhook
		if err := json.Unmarshal(raw, &wh); err == nil {
			msg := extractPayloadText(wh.Data.Message)
			fmt.Printf("[%s] >> sent: %s\n", wh.CreatedAt.Format(time.RFC3339), msg)
		}

	case "message.delivered":
		var wh apiv3.MessageDeliveredWebhook
		if err := json.Unmarshal(raw, &wh); err == nil {
			msgID := "unknown"
			if wh.Data.MessageId != nil {
				msgID = *wh.Data.MessageId
			}
			fmt.Printf("[%s] -- delivered (message: %s)\n", wh.CreatedAt.Format(time.RFC3339), msgID)
		}

	case "message.read":
		var wh apiv3.MessageReadWebhook
		if err := json.Unmarshal(raw, &wh); err == nil {
			msgID := "unknown"
			if wh.Data.MessageId != nil {
				msgID = *wh.Data.MessageId
			}
			fmt.Printf("[%s] -- read (message: %s)\n", wh.CreatedAt.Format(time.RFC3339), msgID)
		}

	case "message.failed":
		var wh apiv3.MessageFailedWebhook
		if err := json.Unmarshal(raw, &wh); err == nil {
			msgID := "unknown"
			if wh.Data.MessageId != nil {
				msgID = *wh.Data.MessageId
			}
			fmt.Printf("[%s] !! failed (message: %s, code: %d)\n", wh.CreatedAt.Format(time.RFC3339), msgID, wh.Data.Code)
		}

	default:
		fmt.Printf("[%s] %s\n", time.Now().Format(time.RFC3339), peek.EventType)
	}
}

func extractPayloadText(msg *apiv3.MessagePayload) string {
	if msg == nil || msg.Parts == nil {
		return ""
	}
	for _, part := range *msg.Parts {
		textPart, err := part.AsSchemasTextPartResponse()
		if err == nil && textPart.Value != "" {
			return textPart.Value
		}
	}
	return "(media)"
}

func init() {
	listenCmd.Flags().Int("port", 4040, "Local port to listen on")
	listenCmd.Flags().Bool("json", false, "Output events as JSON lines")
	listenCmd.Flags().StringSlice("events", nil, "Event types to subscribe to")
	rootCmd.AddCommand(listenCmd)
}

File: src/commands/listen.ts
import { Command, Flags } from '@oclif/core';
import {
  createServer,
  type IncomingMessage,
  type ServerResponse,
} from 'node:http';
import type { components } from '../gen/api-types.js';

type MessageReceivedWebhook = components['schemas']['MessageReceivedWebhook'];
type MessageSentWebhook = components['schemas']['MessageSentWebhook'];
type MessageDeliveredWebhook = components['schemas']['MessageDeliveredWebhook'];
type MessageReadWebhook = components['schemas']['MessageReadWebhook'];
type MessageFailedWebhook = components['schemas']['MessageFailedWebhook'];
type MessagePayload = components['schemas']['MessagePayload'];

interface WebhookEvent {
  event_type?: string;
  created_at?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  data?: any;
}

export default class Listen extends Command {
  static override description = 'Listen for incoming messages';

  static override examples = [
    '<%= config.bin %> <%= command.id %>',
    '<%= config.bin %> <%= command.id %> --port 4567',
    '<%= config.bin %> <%= command.id %> --events message.received,message.delivered',
    '<%= config.bin %> <%= command.id %> --json',
  ];

  static override flags = {
    port: Flags.integer({
      char: 'p',
      description: 'Local port to listen on',
      default: 4040,
    }),
    json: Flags.boolean({
      description: 'Output events as JSON lines',
      default: false,
    }),
    events: Flags.string({
      description: 'Event types to filter (comma-separated)',
    }),
  };

  private server: ReturnType<typeof createServer> | null = null;

  async run(): Promise<void> {
    const { flags } = await this.parse(Listen);

    const eventFilter = flags.events?.split(',').map((e) => e.trim()) || null;

    await new Promise<void>((resolve, reject) => {
      this.server = createServer((req, res) => {
        this.handleRequest(req, res, flags.json, eventFilter);
      });

      this.server.on('error', (err) => {
        reject(err);
      });

      this.server.listen(flags.port, () => {
        this.log(`Listening for events on :${flags.port}`);
        this.log('Press Ctrl+C to stop\n');
      });

      // Handle graceful shutdown
      const shutdown = () => {
        this.log('\nShutting down...');
        this.server?.close(() => {
          resolve();
        });
      };

      process.on('SIGINT', shutdown);
      process.on('SIGTERM', shutdown);
    });
  }

  private handleRequest(
    req: IncomingMessage,
    res: ServerResponse,
    jsonOutput: boolean,
    eventFilter: string[] | null
  ): void {
    if (req.method !== 'POST' || req.url !== '/webhook') {
      res.writeHead(req.url === '/webhook' ? 405 : 404);
      res.end();
      return;
    }

    let body = '';
    req.on('data', (chunk) => {
      body += chunk;
    });

    req.on('end', () => {
      try {
        const event = JSON.parse(body) as WebhookEvent;

        // Apply event filter if specified
        if (
          eventFilter &&
          event.event_type &&
          !eventFilter.includes(event.event_type)
        ) {
          res.writeHead(200);
          res.end();
          return;
        }

        if (jsonOutput) {
          this.log(body);
        } else {
          this.printEvent(event);
        }

        res.writeHead(200);
        res.end();
      } catch {
        res.writeHead(400);
        res.end('Invalid JSON');
      }
    });
  }

  private printEvent(event: WebhookEvent): void {
    const timestamp = event.created_at || new Date().toISOString();

    switch (event.event_type) {
      case 'message.received': {
        const wh = event as MessageReceivedWebhook;
        const from = wh.data?.from || 'unknown';
        const msg = this.extractText(wh.data?.message);
        this.log(`[${timestamp}] << ${from}: ${msg}`);
        break;
      }

      case 'message.sent': {
        const wh = event as MessageSentWebhook;
        const msg = this.extractText(wh.data?.message);
        this.log(`[${timestamp}] >> sent: ${msg}`);
        break;
      }

      case 'message.delivered': {
        const wh = event as MessageDeliveredWebhook;
        const msgId = wh.data?.message_id || 'unknown';
        this.log(`[${timestamp}] -- delivered (message: ${msgId})`);
        break;
      }

      case 'message.read': {
        const wh = event as MessageReadWebhook;
        const msgId = wh.data?.message_id || 'unknown';
        this.log(`[${timestamp}] -- read (message: ${msgId})`);
        break;
      }

      case 'message.failed': {
        const wh = event as MessageFailedWebhook;
        const msgId = wh.data?.message_id || 'unknown';
        const code = wh.data?.code || 'unknown';
        this.log(`[${timestamp}] !! failed (message: ${msgId}, code: ${code})`);
        break;
      }

      default:
        this.log(`[${timestamp}] ${event.event_type || 'unknown'}`);
    }
  }

  private extractText(message: MessagePayload | undefined): string {
    if (!message?.parts) {
      return '';
    }

    for (const part of message.parts) {
      if ('value' in part && typeof part.value === 'string') {
        return part.value;
      }
    }

    return '(media)';
  }
}
